src/hotspot/share/classfile/classLoaderHierarchyDCmd.cpp

```diff
@@ -287,8 +315,9 @@ class LoaderTreeNode : public ResourceObj {
   // and they have the same name or no name (note: leaf check is done by caller).
   bool can_fold_into(LoaderTreeNode* target_node) const {
     assert(is_leaf() && target_node->is_leaf(), "must be leaf");
-    return _cld->class_loader_klass() == target_node->_cld->class_loader_klass() &&
-           _cld->name() == target_node->_cld->name();
+    return (loader_klass() != NULL) && (loader_klass() == target_node->loader_klass()) &&
+           ((loader_name_oop() == NULL && target_node->loader_name_oop() == NULL) ||
+             (::strcmp(loader_name(), target_node->loader_name()) == 0));
```

Comment:
This is hard to parse. I think using if/else statements would be cleaner.

Reply:
Fixed